File: web-next/src/composables/grants.ts
/// Granting libraries to a user: a whole-SET write, optimistic, versioned.
///
/// Every one of those words is a bug that was fixed here.
///
/// **Whole-set**, so the chips a click computes from have to be up to date or
/// the click undoes the last one. They were not — the state only moved when
/// the round trip finished, so granting two libraries in quick succession sent
/// the same pre-click set twice and the second write lost the first. Both
/// reported success. Measured: two grants in, one granted.
///
/// **Optimistic**, so the chips move now and the commits wait. Filtering stale
/// replies does not order the writes: request A can commit after B and leave
/// the hub holding A while the panel shows B.
///
/// **Versioned** (UI-25), so two admins editing the same user cannot silently
/// overwrite each other. The version the next write must carry comes from the
/// last ANSWER, never from the row — the queue means a second click goes out
/// before any repaint, and the row still holds the version the first one
/// consumed.

import { ref } from 'vue'

import { adminSetUserLibraries } from '../api/generated/kahawai.ts'
import { SerialQueue } from './serial.ts'

export type Access = { all_libraries: boolean; libraries: string[] }

type Write = {
  /// Which write is newest.
  seq: number
  /// How many are out.
  inflight: number
  /// Which answer the revert target came from, and what it is.
  savedSeq: number
  saved: Access | null
  /// The `grants_version` the next write must carry.
  version: number
}

export function useGrants(options: {
  /// Paint an access set onto the row.
  show: (userId: string, access: Access) => void
  /// The version the hub has now, onto the row — a different question from
  /// the chips, and it must not paint them.
  version: (userId: string, version: number) => void
  /// Something the operator did, that did not work.
  refused: (why: string) => void
  /// Read the users again. Only for the one refusal that leaves this panel
  /// holding a version the hub has moved past.
  reread: () => void
}) {
  const queue = new SerialQueue()
  /// Per user, because two users' grants are two independent writes and a
  /// shared counter would have them cancel each other.
  const writes = new Map<string, Write>()
  const busy = ref(new Set<string>())

  function setBusy(id: string, out: boolean) {
    const next = new Set(busy.value)
    if (out) next.add(id)
    else next.delete(id)
    busy.value = next
  }

  return {
    busy,
    async set(
      user: { id: string; all_libraries: boolean; libraries: string[]; grants_version: number },
      all: boolean,
      libraries: string[],
    ): Promise<void> {
      const write = writes.get(user.id) ?? {
        seq: 0,
        inflight: 0,
        savedSeq: 0,
        saved: null,
        version: user.grants_version,
      }
      writes.set(user.id, write)

      if (write.inflight === 0) {
        // Nothing outstanding means the chips on screen are what the hub has.
        write.saved = { all_libraries: user.all_libraries, libraries: user.libraries }
        // And the row is the hub's answer — but only if that answer is not
        // OLDER than what is already known. A read that started before a write
        // and lands after it carries the pre-write snapshot, and taking its
        // version back would send a spent one and be told "somebody else
        // changed this" when nobody had. Versions only climb.
        write.version = Math.max(write.version, user.grants_version)
      }

      const mine = ++write.seq
      write.inflight++
      setBusy(user.id, true)
      options.show(user.id, { all_libraries: all, libraries })

      try {
        const answer = await queue.run(user.id, () =>
          adminSetUserLibraries(user.id, {
            all_libraries: all,
            libraries,
            grants_version: write.version,
          }),
        )
        // Whatever else this answer is, it moved the version on — and the
        // queue means the next write is not sent until this lands. Onto the
        // ROW as well: leaving the row holding the version this write just
        // consumed made every second edit send a spent one and come back
        // `stale_write`, telling an operator somebody else had changed it when
        // nobody had.
        write.version = answer.grants_version
        options.version(user.id, answer.grants_version)

        // The revert target moves on ANY success, newest-first: an older write
        // succeeding while a newer one is out still tells us something the hub
        // has accepted, and reverting past it would undo it.
        if (mine > write.savedSeq) {
          write.savedSeq = mine
          write.saved = { all_libraries: answer.all_libraries, libraries: answer.libraries }
        }

        // The CHIPS only ever come from the newest write. Grant A, grant B, A
        // answers first: painting A drops the chips back while B is still out,
        // and the next click there sends [A, C] and revokes B for real.
        if (mine !== write.seq) return
        options.refused('')
        if (write.saved) options.show(user.id, write.saved)
      } catch (cause) {
        // An older write failing says nothing about where the newest one is
        // going, and reverting to what was on screen two clicks ago would undo
        // a grant the operator has since made.
        if (mine === write.seq) {
          if (write.saved) options.show(user.id, write.saved)
          options.refused(String(cause))
        }
        // A refused write leaves this panel holding a version the hub has
        // moved past, and nothing else here would go and look: granting
        // libraries emits no hint. Every further click would send the same
        // spent version and be refused again — on an idle hub, for ever, with
        // a message blaming an admin who was not there. Reading once is what
        // the message asks the operator to do; doing it for them is the whole
        // difference between a refusal and a dead panel.
        if ((cause as { code?: string }).code === 'stale_write') options.reread()
      } finally {
        write.inflight--
        if (write.inflight === 0) setBusy(user.id, false)
      }
    },
  }
}

File: crates/kahawai-hub/src/grants.rs
//! Runtime grants reference external mediadb library IDs. Only the user FK is
//! stored here; library existence, composition and item visibility come from Store.
//! Removing a library makes dangling grants inert because IDs are never reused.
//! Runtime writes validate library IDs through `set_catalogue_access`.
//!
//! Per-library access grants (HUB-10): which libraries an account may
//! see, and so which items it may browse, search, open, fetch artwork
//! for, download subtitles for, and play.
//!
//! ## What is stored
//!
//! `users.all_libraries`, a flag defaulting to 1; `user_libraries`, a plain
//! (user, library) list; and `users.grants_version`, a counter bumped by every
//! write to either of them:
//!
//! | `all_libraries` | rows in `user_libraries` | the account sees |
//! |-----------------|--------------------------|------------------|
//! | 1               | ignored                  | every library, including ones created later |
//! | 0               | some                     | exactly those |
//! | 0               | none                     | nothing |
//!
//! `grants_version` is what makes a wholesale write safe to make concurrently
//! (UI-25). A reader is told the version it read at; a writer sends it back
//! and the `UPDATE` matches only while it still holds, so of two admins who
//! read the same state and both submit, the second is refused rather than
//! silently replacing the first. See [`set_catalogue_access`].
//!
//! The flag is there so that "nothing" is expressible. A list on its own
//! cannot say it: with no rows meaning "everything" you can never revoke
//! the last library, and with no rows meaning "nothing" every library
//! created later is invisible until someone remembers to hand it out.
//! The flag also makes the upgrade honest — an existing hub migrates with
//! every account unrestricted, which is what it had a moment earlier.
//!
//! Grants attach to libraries, never directly to collections (0008): a
//! collection owns physical copies, while a library is the composition a
//! person is given. Catalogue entries are visible through accessible assigned
//! copies. Children require their own accessible copy; album associations also
//! require an accessible supporting copy on that specific album.
//!
//! An item in no library at all is invisible to a restricted account.
//! There is nothing to grant that would reach it; attach its collection
//! to a library first.
//!
//! ## Who bypasses
//!
//! Admins, always — which is why every entry point here takes [`Claims`]
//! rather than a user id, so a caller cannot forget. The admin role is
//! the configuration role: it can grant itself any library through the
//! same endpoint that would restrict it, so enforcing one against it is
//! theatre with a per-request cost. An admin's grant rows are stored and
//! simply not consulted until `is_admin` comes off.
//!
//! ## Why 404 and never 403
//!
//! Denials answer 404, for library ids as much as item ids. 403 says
//! "this exists and is not yours", which turns every endpoint taking an
//! id into an oracle for the rest of the catalogue. 404 tells a
//! restricted account the one thing it is entitled to know: its own
//! grants did not cover that. Callers own the status; what this module
//! returns is a bool.
//!
//! ## Cost
//!
//! An unrestricted account — every account on a single-user hub, and
//! every admin — never reaches the predicates below. Callers resolve
//! [`restricted`] once and otherwise run the SQL they always ran, so the
//! measured NFR-1 browse plans are untouched. A restricted account pays
//! one indexed read per item-scoped request, and one probe per candidate
//! row on the two scan-shaped browses: the cost class of the in-library
//! search predicate that has always been there.

use anyhow::Result;
use kahawai_sqlite::Database as SqlitePool;
use serde::Serialize;
use sqlx::Row;
use utoipa::ToSchema;

use crate::auth::Claims;

/// Whether the grant predicates apply to this request at all.
///
/// An unknown user id reads as restricted-with-no-grants. It cannot
/// happen behind `require_auth` — the token was signed for an account —
/// but the direction a missing row falls is not something to leave to
/// `unwrap_or_default`.
pub async fn restricted(db: &SqlitePool, claims: &Claims) -> Result<bool> {
    if claims.admin {
        return Ok(false);
    }
    let all: Option<i64> = sqlx::query_scalar("SELECT all_libraries FROM users WHERE id = ?")
        .bind(&claims.sub)
        .fetch_optional(db)
        .await?;
    Ok(all.unwrap_or(0) == 0)
}

/// May this account see this library? True for a library that does not
/// exist when the account is unrestricted — "no such library" is the
/// caller's own 404 to give, not a grant decision.
pub async fn can_see_library(db: &SqlitePool, claims: &Claims, library_id: &str) -> Result<bool> {
    if claims.admin {
        return Ok(true);
    }
    let ok: i64 = sqlx::query_scalar(
        "SELECT EXISTS (SELECT 1 FROM users WHERE id = ?1 AND all_libraries = 1)
             OR EXISTS (SELECT 1 FROM user_libraries WHERE user_id = ?1 AND library_id = ?2)",
    )
    .bind(&claims.sub)
    .bind(library_id)
    .fetch_one(db)
    .await?;
    Ok(ok != 0)
}

#[derive(Debug, Serialize, ToSchema)]
pub struct UserAccess {
    pub id: String,
    pub username: String,
    pub is_admin: bool,
    pub all_libraries: bool,
    pub created_at: i64,
    pub libraries: Vec<String>,
    /// What this account's grants were when they were read, for the write
    /// that follows. See [`set_catalogue_access`].
    pub grants_version: i64,
}

/// Every account with its access, for the admin panel. Sorted by name,
/// which is how the panel lists them and how a diff between two hubs
/// stays readable.
pub async fn users_with_access(db: &SqlitePool) -> Result<Vec<UserAccess>> {
    let rows = sqlx::query(
        "SELECT u.id, u.username, u.is_admin, u.all_libraries, u.created_at, u.grants_version,
                (SELECT json_group_array(ul.library_id)
                   FROM user_libraries ul WHERE ul.user_id = u.id) AS libraries
           FROM users u ORDER BY u.username",
    )
    .fetch_all(db)
    .await?;
    Ok(rows
        .iter()
        .map(|r| UserAccess {
            id: r.get("id"),
            username: r.get("username"),
            is_admin: r.get::<i64, _>("is_admin") != 0,
            all_libraries: r.get::<i64, _>("all_libraries") != 0,
            created_at: r.get("created_at"),
            grants_version: r.get("grants_version"),
            libraries: serde_json::from_str(&r.get::<String, _>("libraries")).unwrap_or_default(),
        })
        .collect())
}

/// What a write to an account's grants did. Each is something a caller can
/// act on, so they are return values rather than errors to read prose out of;
/// an `Err` from [`set_catalogue_access`] is the database being unavailable.
#[derive(Debug, PartialEq, Eq)]
pub enum SetAccess {
    Applied {
        grants_version: i64,
        /// What was stored, read inside the same transaction that wrote it.
        ///
        /// The caller used to read it back afterwards, which left a window: a
        /// second admin's write landing in between paired THIS write's version
        /// with THEIR library set, and the panel painted their chips as ours.
        /// It self-healed on the next click — a spent version is refused — but
        /// a feature whose whole purpose is that the two agree should not have
        /// a moment where they do not.
        libraries: Vec<String>,
    },
    /// Somebody else wrote since this admin read. UI-25: the panel sends the
    /// COMPLETE set rather than a delta, so a second write does not merge with
    /// the first — it replaces it, and the first admin's change is gone with
    /// nothing said. A version turns that into a refusal they can see.
    Stale,
    NoSuchUser,
}

async fn write_access(
    db: &SqlitePool,
    user_id: &str,
    expected: i64,
    all_libraries: bool,
    libraries: &[String],
) -> Result<SetAccess> {
    let mut tx = db.begin().await?;
    let res = sqlx::query(
        "UPDATE users SET all_libraries = ?, grants_version = grants_version + 1
          WHERE id = ? AND grants_version = ?",
    )
    .bind(all_libraries)
    .bind(user_id)
    .bind(expected)
    .execute(&mut *tx)
    .await?;
    if res.rows_affected() == 0 {
        // Absent, or written since. Tell them apart so an admin looking at a
        // user who was deleted under them does not read "try again".
        let exists: Option<i64> = sqlx::query_scalar("SELECT 1 FROM users WHERE id = ?")
            .bind(user_id)
            .fetch_optional(&mut *tx)
            .await?;
        return Ok(if exists.is_some() {
            SetAccess::Stale
        } else {
            SetAccess::NoSuchUser
        });
    }
    sqlx::query("DELETE FROM user_libraries WHERE user_id = ?")
        .bind(user_id)
        .execute(&mut *tx)
        .await?;
    for library_id in libraries {
        sqlx::query("INSERT OR IGNORE INTO user_libraries (user_id, library_id) VALUES (?1,?2)")
            .bind(user_id)
            .bind(library_id)
            .execute(&mut *tx)
            .await?;
    }
    let grants_version: i64 = sqlx::query_scalar("SELECT grants_version FROM users WHERE id = ?")
        .bind(user_id)
        .fetch_one(&mut *tx)
        .await?;
    let stored: Vec<String> =
        sqlx::query_scalar("SELECT library_id FROM user_libraries WHERE user_id = ?")
            .bind(user_id)
            .fetch_all(&mut *tx)
            .await?;
    tx.commit().await?;
    tracing::info!(
        user_id,
        all_libraries,
        granted = libraries.len(),
        grants_version,
        "library access set"
    );
    Ok(SetAccess::Applied {
        grants_version,
        libraries: stored,
    })
}

/// External mediadb library IDs are validated before entering the hub transaction.
/// A concurrent library deletion can leave a dangling reference; it grants
/// nothing, because visibility always requires a current mediadb library.
pub async fn set_catalogue_access(
    db: &SqlitePool,
    store: &kahawai_mediadb::Store,
    user_id: &str,
    expected: i64,
    all_libraries: bool,
    libraries: &[String],
) -> Result<SetAccess> {
    let existing = store.libraries().await?;
    let valid: Vec<String> = libraries
        .iter()
        .filter(|id| existing.iter().any(|l| l.id == **id))
        .cloned()
        .collect();
    write_access(db, user_id, expected, all_libraries, &valid).await
}
